Implement new textobject for indentation level

EpocSquadron opened the pull request:
This implements a tetobject corresponding to the current indentation
level of the selection. It is only implemented for the "match mode"
and has no difference between inside and around (as there's no obvious
meaning) to those. It does however take a count, where count extends
the selection to "lower" levels of indentation than represented by the
current selection.

I couldn't see a sensible way to do "next" and "prev" by this
textobject. Although currently it will extend past empty lines, so
maybe it could be adjusted to consider "next" and "prev" to be the next
contiguous block of the same indentation level as the current selection,
after one or more blank lines. That may also allow for defining "around"
as the current behaviour, and use "inside" to stop at blank lines. Open
to thoughts.

EpocSquadron commented:
I've updated the original description to hopefully be more readable, included a screencast of the feature, and made clear the call for feedback on potentially extending the logic.

daedroza commented:
Hey @EpocSquadron , this look like a great change, for next and prev implementation, only the contiguous block should matter. I would be waiting for that change if you're willing to implement, thanks again!

shaleh commented on helix-core/src/textobject.rs:
```diff
@@ -198,6 +200,83 @@ pub fn textobject_paragraph(
+        return range;
+    }
+
+    min_indent = Some(min_indent.unwrap() + 1 - count);
```
This could be a redefinition of `min_indent` as a number since from this point on it is always Some. That removes any need for unwrap below.

EpocSquadron commented:
I've updated this branch, squashed the changes, addressed feedback, and updated the description of the PR to reflect the current state. I removed the short screen recording I had up prior as it no longer accurately reflected the functionality. I ran out of time to record a new one.

@daedroza I'm not able to dedicate time towards the next/prev motions at the moment. I may return at a later date but no promises or timeline. I did at least move forward with distinguishing the inside and around versions, so hopefully that is sufficient to get this merged and expand on later.

crasm commented:
Hello, just wanted to register my support for merging this feature. I like being able to do this kind of selection to be able to easily sort lists and arguments in my nix config. It's one of the biggest things missing in helix that I'm used to in neovim.

yerlaser commented:
>I could implement "next" (]i) and "prev" ([i) to be the next contiguous block of the same indentation level as the current selection, after one or more blank lines. However I don't want to hold this implementation back as it is now complete and I have little time to dedicate to further features. Others are welcome to implement, or I may return to it at a later date.

This has been implemented in https://github.com/helix-editor/helix/pull/9843
